### PR TITLE
Improving directories in playlists

### PR DESCRIPTION
**Pull request by christoph-heinrich**
Adds a new option `--directory-mode=recursive|lazy|ignore` to control recursive loading of directories, as suggested in https://github.com/mpv-player/mpv/pull/11292#issuecomment-1426792272

Sort files in front of directories, as suggested in https://github.com/mpv-player/mpv/pull/11292#issuecomment-1426701499

### CODE REVIEW

**Comment by Dudemanguy** [comment]
Just at a quick glance, you could probably just put `demux_playlist_opts` inside `pl_parser` and save yourself a lot of trouble.

**Comment by christoph-heinrich** [comment]
Thanks for the tip, while changing that I also realized that I could add to the playlist directly instead of that `files` array.

**Comment by christoph-heinrich** [comment]
I've implemented recursion in [a branch](https://github.com/christoph-heinrich/mpv/blob/playlist_directory_next/TOOLS/lua/autoload.lua) in case someone wants to try it.

**Comment by christoph-heinrich** [comment]
I've brought the new option in line with how it's done in all demuxers except for demux_libarchive.

The mechanism that demux_libarchive uses was added in 8150b8552b4ed1b4c33089bdb1f9159b181660ab and 5016a1e4a6118e1b91b1ee1ca5221ae09cf8753e, so it looks like wm4 wanted take things in that direction.

**Comment by Dudemanguy** [comment]
Yeah going through `OPT_BOOL` stuff, I noticed that wm4 added those commits. Honestly, I'm not really sure what the purpose is.

File: demux/demux_playlist.c [hunk 1]
```diff
@@ -309,10 +327,27 @@ static bool same_st(struct stat *st1, struct stat *st2)
     return st1->st_dev == st2->st_dev && st1->st_ino == st2->st_ino;
 }
 
+struct pl_dir_entry {
+    char* path;
+    char* name;
```
Comment (Dudemanguy) [inline demux/demux_playlist.c:332]:
```suggestion
    char *path;
    char *name;
```

**Review by Dudemanguy — APPROVED** [review]
I like this. It's a good feature, and I agree with sorting directories after the files in the toplevel. It makes the most sense.

File: DOCS/man/options.rst [hunk 2]
```diff
@@ -3937,6 +3937,13 @@ Demuxer
     libarchive opens all volumes anyway when playing the main file, even though
     mpv iterated no archive entries yet.
 
+``--demuxer-playlist-recursive-dir=<yes|no>``
```
Comment (Dudemanguy) [inline DOCS/man/options.rst:3940]: Just had a thought, but how about renaming this to `recursive-dir` for simplicity?

Comment (christoph-heinrich) [reply DOCS/man/options.rst:3940]: We could do that, as the name is rather long, but the other demuxer specific options also have `demuxer-<demuxer name>-<option>`.

Comment (Dudemanguy) [reply DOCS/man/options.rst:3940]: `sub-codepage` and `rtsp-transport` in demux_lavf don't, so there's precedent.

**Comment by Dudemanguy** [comment]
I know I'm adding another request out of nowhere, but what about making this a threeway choice? recursive, open the directories when you get to them, and finally ignoring all directories (i.e. only play the top level files)?

**Comment by christoph-heinrich** [comment]
It's `--directory-mode=recursive|lazy|ignore` now.

File: DOCS/man/options.rst [hunk 3]
```diff
@@ -3937,6 +3937,13 @@ Demuxer
     libarchive opens all volumes anyway when playing the main file, even though
     mpv iterated no archive entries yet.
 
+``--directory-mode=<recursive|lazy|ignore>``
+    When opening a directory, open subdirectories recursively, lazily or not at
+    all (default: yes).
```
Comment (Dudemanguy) [inline DOCS/man/options.rst:3942]: `default: recursive`

File: DOCS/man/options.rst [hunk 4]
```diff
@@ -3937,6 +3937,13 @@ Demuxer
     libarchive opens all volumes anyway when playing the main file, even though
     mpv iterated no archive entries yet.
 
+``--directory-mode=<recursive|lazy|ignore>``
+    When opening a directory, open subdirectories recursively, lazily or not at
+    all (default: recursive).
+
+    Disabling this can lead to problems with resuming playlists
+    (`RESUMING PLAYBACK`_) and possibly other things.
```
Comment (Dudemanguy) [inline DOCS/man/options.rst:3945]: Sorry I guess this should probably be reworded too. How about, `Values other than ``recursive`` can lead to problems...`

**Comment by garoto** [comment]
> It's `--directory-mode=recursive|lazy|ignore` now.

I like the option name and its options now. Very mpv-like. Fricking neat.

File: demux/demux_playlist.c [hunk 5]
```diff
@@ -35,6 +36,31 @@
...
+        {"directory-mode", OPT_CHOICE(dir_mode,
+            {"recursive", DIR_RECURSIVE},
+            {"lazy", DIR_LAZY},
+            {"ignore", DIR_IGNORE})},
+        {0}
+    },
+    .size = sizeof(struct demux_playlist_opts),
+    .defaults = &(const struct demux_playlist_opts){
+        .dir_mode = 0
+    },
```
Comment (Dudemanguy) [inline demux/demux_playlist.c:61]:
```suggestion
    .defaults = &(const struct demux_playlist_opts) {0},
```

For consistency mainly.

**Pull request merged**